### ISSUE TITLE
Bug on scrolling to top

### ISSUE THREAD

**Issue by stereobooster** [civiccc/react-waypoint#202]
I have list of numbered items. Waypoint seted up like this

```jsx
  <Waypoint onEnter={ props.onEnter.bind(null, props.place) } bottomOffset="70%" topOffset="10%">
    <div><Card {...props} /></div>
  </Waypoint>

...

onEnterHandler(e) {
    console.log(e.number)
}
```

When I scroll from top to bottom I see (in console)

```
1
2
3
4
5
6
7
8
```

But when scroll from bottom to top

```
11
12
6
7
8
9
10
3
4
5
1
2
```

PS you need to scroll really fast to reproduce bug

I will try to craft example. Is this known thing? Or I did something strange to waypoint.

**Comment by jmeas** [comment]
Hi @stereobooster , sorry that you're having issues with Waypoint 🙁 . We appreciate that you opened this issue, though! I have some questions that could help me debug this:

1. which version of Waypoint are you using?
2. how much space is between your waypoints?
3. do you have an example on GitHub that we could check out?

Thanks!

**Comment by stereobooster** [comment]
1. `"react-waypoint": "^7.0.3", "react": "^15.5.4",`
2. `24px`
3.  I owe you this one

**Comment by stereobooster** [comment]
@jmeas Here you go https://github.com/stereobooster/waypoint-bug-report

**Comment by jmeas** [comment]
@stereobooster , thank you! I'll take a look at this tonight :)

**Comment by jmeas** [comment]
Just thought I'd post an update. I was able to clone down this repository and reproduce the behavior that you're seeing. I'll need to dig more into it to see what the cause is. I may have time tonight, but it'll definitely be this week.

Thanks _a lot_ for making this repository, @stereobooster ! It'll be a huge help to getting to the bottom of this. :v:

**Comment by jmeas** [comment]
I've been thinking about this, and I have an explanation that could be correct. I believe that these two facts cause this issue:

1. React renders components from top-to-bottom. In order words, in a chunk of JSX, the things at the top will have their lifecycle methods called before the things at the bottom
2. When you scroll very fast in any direction, two or more Waypoints may get updated within the same render
3. The lifecycle methods of a Waypoint are where we trigger the callbacks

If your waypoints are in your JSX like this:

```
1
2
3
4
5
6
7
```

Then, when scrolling down, they will _always_ be in the right order. Let's assume that within 1 render, we jump from 1 and 2 being onscreen to 5 and 6 being onscreen. React needs to render both of those, and so it does, in this order:

5, and then 6.

This lines up with the direction that we are scrolling.

If we reverse this, we can see how the bug would come into play. Within 1 render, jump from 6 and 7 being onscreen to 2 and 3 being onscreen. React renders them like this:

2, and then 3.

So while your expectation is that the onEnter callback of 3 would be triggered before 2, **this is not guaranteed.**

I need to do more testing to confirm this, but it seems like a reasonable explanation. What do you think, @stereobooster ?

**Comment by stereobooster** [comment]
I suppose this is something like this, buuut you are talking about render. Aren't all elements rendered at this point, isn't it about simply calling `onEnter` callback

**Comment by jmeas** [comment]
@stereobooster , this will affect the entire lifecycle of the component – `render`, `onEnter` (which is triggered from within a lifecycle method, `componentDidUpdate`), and all of the other methods, too.

I just chatted with my coworker who confirmed that this is the expected behavior from React, so I don't think we'll be able to make any changes to Waypoint to account for this. It's a limitation of React itself (and probably any VDom implementation, honestly).

Is this causing a bug in your app?

**Comment by stereobooster** [comment]
> Is this causing a bug in your app?

Yes. Maybe there is workaround? When waypoint enters in the hot-zone on left, I highlight marker on the map on the right. Basically it works light bootstrap scrollspy. Maybe there is other component which I should use?

**Comment by jmeas** [comment]
> When waypoint enters in the hot-zone on left, I highlight marker on the map on the right.

This makes sense. I have this same problem in my own app, now that I think about it (I use Waypoints for scrollspies as well).

> Maybe there is other component which I should use?

**Comment by stereobooster** [comment]
Maybe we can summon someone from React team? They are generally responsive and helpful. cc @gaearon

**Comment by jmeas** [comment]
It's worthwhile getting their input, but I doubt we'll see a fix from the React side. It'd require a pretty big update to the React rendering engine. They're currently working on a big update, "Fiber," and my current understanding is that it doesn't factor in scroll direction.

I like the idea of virtualization libs factoring in scroll direction when it comes to lifecycle methods, but it's a big ask, and I doubt we'll see it.

The only solutions I can think of right now involve going outside of the lifecycle methods of React. For instance, if your items are evenly spaced, then you can use the scroll position of the page to determine which item should be highlighted.

For things that aren't evenly spaced, you can do the same approach, but it requires more work. You'd need to keep track of either the position of the elements, or their height (using refs with React), then use that data in conjunction with the scroll position to figure out which is "active."

There may be simpler solutions, but they elude me right now.

**Comment by jmeas** [comment]
I'm going to close this issue for now because **this is a limitation of creating scrollspies with any and all virtual DOM implementations that I know of**. They all render from top to bottom, and don't keep track of a rendering direction.

I'm pretty sure that there's no way for React Waypoint to solve this issue in a generic way, nor any other React component that attempts to be a scrollspy.

Future visitors to this issue: you may be much more clever than me, and come up with a solution. If you do, please post it here! I'm keen on being proved wrong here.

**Comment by stereobooster** [comment]
Here is how you can hack around https://github.com/stereobooster/waypoint-bug-report/commit/8d108d64c34d970aaf2f8f1a4dd21ec9e04aab74

**Comment by jmeas** [comment]
Nice work, @stereobooster ! It may be worth adding somethin' like this into the core lib if there's a way! If you had time to code it up, I'd be happy to review the work.

**Comment by stereobooster** [comment]
As I said this is hack, not a full featured solution. It totally works in my case (fix the bug), but it comes with drawbacks:
- it is guaranteed that when the user stops scrolling their will get proper state of application
- but it is not guaranteed that intermediate events will be fired in proper order

How it works:
1. detect direction of scrolling with the help of event listener, previous position and current position
2. if it goes bottom - store max number of item in the list as *current item*. Values should increase if you scroll down.
3. if it goes up - store min number of item in the list as *current item*. Values should decrease if you scroll up.

*Current item* is the outcome of this hack. This is exactly what you need for scrollspy. But events are still fired in the wrong order.

More general solution, which can be incorporated in the lib would require to buffer events if they are in wrong order and then fire them in proper order. To do this we will need to introduce debounce function, which will introduce some latency in event firing process. So the question if this the solution which you expect and will accept?

**Comment by jmeas** [comment]
I think it’s worthwhile to at least consider. I’m curious to get @trotzig / @lencioni ‘s thoughts.

—

tl;dr: React emits lifecycle methods for any given set of elements from top to bottom, independent of scroll direction. So if you scroll from the bottom of the page to the top of the page quickly, such that 3 waypoints appear onscreen at once, you’d expect them to fire lifecycle methods (including `onEnter`, etc.) in this order:

3, 2, 1

yet they will fire in this order:

1, 2, 3

due to React starting its lifecycle at the top-most Waypoint.

@stereobooster is suggesting that we track the scroll direction in Waypoint, then debounce our lifecycle methods a little bit to make sure we can reorder them in situations when the user is scrolling too fast.

I suppose this could be opt-in via a prop. What do you two think, @lencioni / @trotzig ? I tested this on the apps we’re using this scrollspy on, and I can reproduce this unexpected behavior pretty easily with a fast scroll.

**Comment by trotzig** [comment]
I think this is definitely worth exploring.

**Comment by maxkoshel** [comment]
I have the same problem and have noticed that sometimes when you scroll very fast some events may be missed or can go in the wrong order. But I have found a fast hack. If you have some space between waypoints (for example 20-30px), set offsets to big value, for example 49% for top and bottom if you want to detect waypoint change in the center of viewport or any other proportions (leave a narrow gap in 1-2%).

**Comment by leanne1** [comment]
I have a similar problem. I have a set of waypoints that fire in a given order say 1, 2, 3, 4 as you scroll down the container. Then i have a button that scrolls the scrollable view back upwards to a given position by setting `element.scrollTop` on the scrollable container. Since the container is effectively scrolling in reverse when this button is clicked I would expect the waypoints to fire in reverse as they do with manual upwards scrolling e.g. 4, 3 ,2, 1, but they don't they fire in the same order as if i had scrolled down 1, 2, 3, 4. So i would say in its current form automated scrolling is not supported by the library. Note my auto scrolling is not animated, i don't know if animating it would fix this i assume it would but that is a constraint that does not work for my project.

**Comment by gennbo** [comment]
Does anyone have a solution or a workaround for this behaviour?
I've stumbled upon this issue when doing a page with multiple sections with own IDs and waypoints and noticed that those IDs are not firing up in correct order when navigating through hash links

**Comment by trotzig** [comment]
I don't know of a good solution or a workaround right now, but I'd be happy to review a PR exploring this. One quick solution that could be worth exploring (bear with me, I haven't thought this out completely) is to iterate backwards through scroll listeners when scrolling up. In most cases, that could be sufficient. I believe that would mean either modifying the [consolidated-events lib](https://github.com/lencioni/consolidated-events) or roll our own scroll event handler consolidation.

**Comment by Resin01** [comment]
Has anyone found a proper solution for this? Or is there one on the way?
